### DISCUSSION TITLE
How Clickhouse deletes data/parts if Order by doesn't have time as first column?

### DISCUSSION THREAD

**Discussion by Sourish-Mukherjee** [ClickHouse/ClickHouse#72839] [Q&A] [answered]
Hi, Need some clarification, on deletion of data by TTL.
 
 This is my table configuration right now. Now I went throught clickhouse docs and learnt that in order by we should put LowCardinality first and then highcardinality. Currently you can consider we have around 500K unique time_stamp, and 1K user_id. Because of which whenever we query, if we query with time_stamp and user_id, or just time_stamp, the granules selected is more or less same. (This we understood because of algorithm works, and since time_stamp is highcardinality), even though i have user_id as filter, granules are not lowering.
 
 Old Schema:
 
 `PARTITION BY toStartOfWeek(toDateTime(time_stamp)) ORDER BY (toDateTime(time_stamp), user_id) TTL toStartOfWeek(toDateTime(time_stamp)) + toIntervalWeek(60) SETTINGS index_granularity = 8192`
 
 To fix this, we have changed the order_by where user_id becomes the first_col of order_by and then time_stamp. This has solved the problem where with user_id and time_stamp, the number of granules getting selected is very less compared to just time.
 
 But, i have few question here:
  1. Will this impact how Clickhouse deletes data in background with TTL. As the order_by is now based on user_id, then time_stamp, wont clickhouse have to scan the entire rows in parts to figure out which rows to delete ? If not why ? How clickhouse chooses rows / parts to delete (couldnt find any doc)
  2. What exactly is ttl_only_drop_parts ? Will this help here if i enable this ? Does it prevent me from scanning entire rows 
 
 New Schema:
 
` PARTITION BY toStartOfWeek(toDateTime(time_stamp)) ORDER BY (toDateTime(time_stamp), user_id) TTL toStartOfWeek(toDateTime(time_stamp)) + toIntervalWeek(60) SETTINGS index_granularity = 8192`

**Comment by den-crane** [comment]
>Now I went throught clickhouse docs and learnt that in order by we should put LowCardinality first and then highcardinality.

It's a general recommendation. It can be sub-optimal in specific applications.

>we have changed the order_by where user_id becomes
>Will this impact how Clickhouse deletes data in background with TTL. 

It does not impact TTL. TTL is not related to primary key / orderby anyhow.

Parts (pieces of the table) have a special attribute `ttl_max_min`, you can check it in the table `system.parts` - `delete_ttl_info_min` /`delete_ttl_info_max` columns. This attribute is calculated during inserts and merges.

TTL merger process starts every 4 hours (by default) and checks this attribute then makes a decision which parts and how should processed (drop or read and remerge).

In your case partitions are aligned with the ttl rule (toStartOfWeek):

```
PARTITION BY toStartOfWeek(toDateTime(time_stamp))

TTL          toStartOfWeek(toDateTime(time_stamp))
```
(TTL is aligned and all rows will expire at the begging of a week) (`delete_ttl_info_min` == `delete_ttl_info_max` always)

So when a whole partition (all rows in a part) are expired simultaneously, then TTL merger just drops the part without a reading the data.
But if the TTL rule is not aligned with a partitioning, then the TTL merge reads the part's rows and creates a new part with not expired rows, it obviously more compute-heavy process.

`ttl_only_drop_parts` -- forbids to the TTL merger to run heavy compute TTL merges and makes the merger to postpone part's TTL processing until all rows in a part are expired.

In your case ttl_only_drop_parts is excessive because TTL rule is aligned with table partitioning. But you can use `ttl_only_drop_parts` -- there is no harm.

----

TLDR: ORDERBY/PRIMARYKEY are not related and don't participate in TTL processing.

**Reply by Sourish-Mukherjee** [reply]
So correct me if I am wrong, 
If my Partitioning logic doesn't match with TTL logic (as in if partition is by `startOfWeek` and TTL is `toDate`), then Clickhouse can't fully utilize the ttl_min and ttl_max metadata, so it will scan through the rows inside part?

If my understanding is clear, TTL doesn't delete any rows, rather when new merge happens, it will marks rows for deletion. Clickhouse then make new parts out of the rows which are not deletable. Because of which it becomes a heavy IO. (So if i have 300 rows in parts, out of which 280 is deleteable, then Clickhouse will create a new part for the remaining 20).

TTL_only_drop_parts save this by preventing this operation. It will force the clickhouse to delete parts only when all rows are expired, thus preventing new parts getting created. So turning it on is actually better for IO right?

Only downside is some rows might be outdated as it's not deleted.

Does clickhouse use the metadata you specified to determine whether a part should be selected, only. When will it only go inside scan rows? During this scanning won't order by matter ? If it does, will it be trivial (now that my time is not the first col of order_by)

**Reply by den-crane** [reply]
>If my Partitioning logic doesn't match with TTL logic (as in if partition is by startOfWeek and TTL is toDate), then Clickhouse can't fully utilize the ttl_min and ttl_max metadata, so it will scan through the rows inside part?

So when a whole partition (all rows in a part) are expired simultaneously, then TTL merger just drops the part without a reading the data.

But if the TTL rule is not aligned with a partitioning, then the TTL merge reads the part's rows and **creates a new part with not expired rows**, it obviously more compute-heavy process.

**Reply by den-crane** [reply, marked as answer]
>Does clickhouse use the metadata you specified to determine whether a part should be selected, only. When will it only go inside scan rows? During this scanning won't order by matter ? If it does, will it be trivial (now that my time is not the first col of order_by)

NO!
Clickhouse reads all rows when it merges the part (TTL or not TTL). 
ORDERBY/PRIMARYKEY are not related and don't participate in TTL processing no matter what.

Today is 2024-12-06 and one part has delete_ttl_info_min = 2024-12-01 and delete_ttl_info_max = 2024-12-31
So bunch of the rows should be deleted because 2024-12-01 < 2024-12-06
Clickhouse merger reads ALL ROWS from this part and writes ALL not expired rows (with date >2024-12-06) to a new part. After that it drop the old part and attaches the new part. No indexes involved here.
